### PR TITLE
SpinLock.TryEnter fail fast for timeout 0

### PR DESCRIPTION
**Pull request by benaadams**
Passes corefx tests

1M iters (single thread, uncontended but locked) for [code](https://gist.github.com/benaadams/a8de9032ae5f6ff89370842fb495d5fa)

```csharp
bool lockTaken = false;
var s = new SpinLock(false);
s.Enter(ref lockTaken);
```

method | pre (ms) | post (ms) | improvement
-------- |------- |------- |------ |
s.TryEnter(0, ref lockTaken) | 24.55 | 6.09 | x 4.0

Need to do the threadpool timing tests I have (on a couple environments), as TrySteal uses this path; and they introduce parallel contention. Hopefully will improve, but will see.

Also corrected some strange spellings.

### CODE REVIEW

File: src/mscorlib/src/System/Threading/SpinLock.cs [hunk 1]
```diff
@@ -336,27 +339,40 @@ private void ContinueTryEnter(int millisecondsTimeout, ref bool lockTaken)
...
                 Thread.EndCriticalRegion();
 #endif
             }
-            else //failed to acquire the lock,then try to update the waiters. If the waiters count reached the maximum, jsut break the loop to avoid overflow
+            else
             {
+                // Lock owned, haven't acquired the lock
+                if (millisecondsTimeout == 0)
+                {
+                    // If timeout is zero fail fast, don't update waiters, don't spin to attempt acquire, exit
```
Comment (stephentoub) [inline src/mscorlib/src/System/Threading/SpinLock.cs]: Nit: missing comma after "If timeout is zero"

**Comment by benaadams** [comment]
Closing for now while I work through the flow a bit more.

**Comment by benaadams** [comment]
@stephentoub I think this logic is clearer? Use a lower watermark rather than a cut off at the high end which simplifies `DecrementWaiters`

Need to do some testing an measurements,

**Comment by benaadams** [comment]
Logic is wrong somewhere

**Pull request closed without merging**